Code/Framework/AzCore/AzCore/Serialization/Json/MapSerializer.cpp

```diff
@@ -231,8 +231,29 @@ namespace AZ
             return context.Report(keyResult, "Failed to read key for associative container.");
         }
 
+        void* valueAddress = nullptr;
+        bool keyExists = false;
+        // Check if we should attempt to update the value if a key already exists
+        if (context.ShouldUpdateAssociativeContainer())
+        {
+            auto associativeContainer = container->GetAssociativeContainerInterface();
+            void* existingKeyValuePair = associativeContainer->GetElementByKey(outputValue, keyElement, keyAddress);
+            if (existingKeyValuePair)
+            {
+                valueAddress = pairContainer->GetElementByIndex(existingKeyValuePair, pairElement, 1);
+                expectedSize--;
+                keyExists = true;
+            }
+        }
+
+        // Either we don't want to update values if the key already exists,
+        // or the key does not exist
+        if (!keyExists)
+        {
+            valueAddress = pairContainer->GetElementByIndex(address, pairElement, 1);
+        }
```

Comment:
It seems like we can defer reserving an element in the Associative container until now, instead of at the beginning of the function.

Reply:
We need to reserve the element in the container earlier in order to read the key(on line 227). The keyAddress points to the first value of the pair reserved for the new element(line 219).

Reply:
Ah I see, in order to process the key in C++, we need it loaded from JSON into the key address